safenode/src/protocol/online_transfers/mod.rs

```diff
@@ -0,0 +1,389 @@
+/// new dbcs (and a change dbc if any), which are returned from this function.
+/// This does not register the transaction in the network.
+/// To do that, the `signed_spends` of each new dbc, has to be uploaded
+/// to the network. When those same signed spends can be retrieved from
+/// enough peers in the network, the transaction will be completed.
+#[allow(clippy::result_large_err)]
+fn crate_transfer_with(selected_inputs: Inputs) -> Result<Outputs> {
+    let Inputs {
+        dbcs_to_spend,
+        recipients,
+        change: (change, change_to),
+    } = selected_inputs;
+
+    let mut inputs = vec![];
+    for (dbc, derived_key) in dbcs_to_spend {
+        let revealed_amount = match dbc.revealed_amount(&derived_key) {
+            Ok(amount) => amount,
+            Err(err) => {
+                warn!("Ignoring dbc, as it didn't have the correct derived key: {err}");
+                continue;
+            }
+        };
+        let input = InputHistory {
+            input: RevealedInput::new(derived_key, revealed_amount),
+            input_src_tx: dbc.src_tx,
+        };
+        inputs.push(input);
+    }
+
+    let mut tx_builder = TransactionBuilder::default()
+        .add_inputs(inputs)
+        .add_outputs(recipients);
+
+    let mut rng = rng::thread_rng();
+
+    let dbc_id_src = change_to.random_dbc_id_src(&mut rng);
+    let change_id = dbc_id_src.dbc_id();
+    if change.as_nano() > 0 {
+        tx_builder = tx_builder.add_output(change, dbc_id_src);
+    }
+
+    // Finalize the tx builder to get the dbc builder.
+    let dbc_builder = tx_builder
+        .build(Hash::default(), &mut rng)
+        .map_err(Error::Dbcs)?;
+
+    // Perform validations of input tx and signed spends,
+    // as well as building the output dbcs.
+    let mut created_dbcs: Vec<_> = dbc_builder
+        .build()
+        .map_err(Error::Dbcs)?
+        .into_iter()
+        .map(|(dbc, amount)| CreatedDbc { dbc, amount })
+        .collect();
+
+    let mut change_dbc = None;
+    created_dbcs.retain(|created| {
+        if created.dbc.id() == change_id {
+            change_dbc = Some(created.dbc.clone());
+            false
+        } else {
+            true
+        }
+    });
+
+    Ok(Outputs {
+        created_dbcs,
+        change_dbc,
+    })
+}
+
+async fn get_fees(dbc_id: DbcId, client: &Client) -> Result<BTreeMap<NodeId, RequiredFee>> {
+    let request = Request::Query(Query::Spend(SpendQuery::GetFees {
+        dbc_id,
+        priority: SpendPriority::Normal,
+    }));
+    let responses = client
+        .send_to_closest(request)
+        .await
+        .map_err(|e| Error::CouldNotGetFees(e.to_string()))?;
+
+    // We just want to receive at least a majority of results, we don't care about any errors
+    // so we log them, but return whatever results we get. If not enough for upper layer, it will error there.
+    let results: BTreeMap<_, _> = responses
+        .into_iter()
+        .flat_map(|res| {
```

Comment:
this actually can be covered by the following `filter_map` ?

Reply:
Then we can't log for the specific type of err. See the logs are different:

flat_map:
`warn!("Error when querying for fees: {error}");`

filter_map:
`warn!("Unexpected response to fee query: {other:?}");`

Reply:
The first flat map removes connection/network type of errors.

The first filter map removes non-query responses (and logs that as an unexpected response, which it is).

The second filter map identifies an error at node within handling of get fees, and logs that. It also identifies an unexpected query response type, and logs that.